File: C#/Question2/NumberFinder.cs
using System.Collections.Generic;

namespace Question2
{
    public static class NumberFinder
    {
        private static readonly HashSet<long> targetNumbers = new HashSet<long>();
        public static long GetNumberOfIndex(int targetIndex)
        {
            // A set that contains all the numbers that satisfy the given conditions
            // HashSet<long> targetNumbers = new HashSet<long>();
            // 1 is a specific case that satisfies the condition, so we add it manually
            targetNumbers.Add(1);
            // Since there is already one element in our targetNumbers set, we set the numberCount to 1
            int numberCount = 1;
            // And start by checking number 2
            long numberToCheckNext = 2;
    
            while (numberCount<targetIndex)
            {
                if (CheckNumber(numberToCheckNext))
                {
                    targetNumbers.Add(numberToCheckNext);
                    numberCount++;
                }
    
                numberToCheckNext++;
            }
    
            return numberToCheckNext-1;
        }
                
        private static bool CheckNumber(long numberToCheckNext)
        {
            int score = 0;
            if (numberToCheckNext % 2 == 0)
            {
                if (targetNumbers.Contains(numberToCheckNext / 2))
                {
                    score += 1;
                }
            }
            if (numberToCheckNext % 3 == 0)
            {
                if (targetNumbers.Contains(numberToCheckNext / 3))
                {
                    score += 1;
                }
            }if (numberToCheckNext % 5 == 0)
            {
                if (targetNumbers.Contains(numberToCheckNext / 5))
                {
                    score += 1;
                }
            }
    
            return score >= 1;
    
        }
    }
}

File: C++/Question2/NumberFinder.cpp
//
// Created by galax on 3/17/2020.
//

#include "NumberFinder.h"

long NumberFinder::GetNumberOfIndex(int targetIndex) {
    std::unordered_set<long> numbers;
    numbers.insert(1);
    int numberCount = 1;
    long numberToCheckNext = 2;

    while(numberCount<targetIndex){
        if(CheckNumber(numberToCheckNext, numbers)){
            numbers.insert(numberToCheckNext);
            numberCount++;
        }
        numberToCheckNext++;
    }
    return numberToCheckNext-1;
}

bool NumberFinder::CheckNumber(long numberToCheckNext, std::unordered_set<long>& targetNumbers) {
    int score = 0;
    if(numberToCheckNext % 2 == 0){
        if(targetNumbers.find(numberToCheckNext/2)!=targetNumbers.end()){
            score++;
        }
    }
    if(numberToCheckNext % 3 == 0){
        if(targetNumbers.find(numberToCheckNext/3)!=targetNumbers.end()){
            score++;
        }
    }
    if(numberToCheckNext % 5 == 0){
        if(targetNumbers.find(numberToCheckNext/5)!=targetNumbers.end()){
            score++;
        }
    }
    return score>=1;
}
